getShifted/getUnshifted methods for BoutOutputs

johnomotani opened the pull request:
Add methods to BoutOutputs class to shift fields orthogonal->field-aligned or the reverse.

Make 'evolvingVariables' a member variable rather than a method.

Add support to DataFile for file-level attributes.

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -516,6 +516,9 @@ class BoutOutputs(object):
 
+
+    :ivar attributes: dictionary of attributes from the first DataFile
+    :ivar evolvingVariables: list of time-evolving variables
```
Please could you put these in [numpydoc][1] style? These can go under an `Attributes` section (see link)

[1]: https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -554,10 +557,15 @@ def __init__(self, path=".", prefix="BOUT.dmp", suffix=None, caching=False,
+            try:
+                npes = f.read("NXPE")*f.read("NYPE")
+            except KeyError:
+                npes = self._attributes["NXPE"]*self._attributes["NYPE"]
```
This is `_attributes`, but `attributes` above. Are they supposed to be different variables?
Also, are `NXPE/NYPE` actually file attributes rather than variables?

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -599,11 +607,47 @@ def keys(self):
+                nz = self["MZ"]
+            except KeyError:
+                try:
+                    nz = self.attributes("mz")
```
I think you meant to use square brackets here

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -599,11 +607,47 @@ def keys(self):
+                try:
+                    nz = self.attributes("mz")
+                except KeyError:
+                    nz = self.attributes["mesh:nz"]
```
I'm a bit confused about where this is coming from -- `mesh:nz` looks like an input file variable, but this class reads output files

ZedThree commented on tools/pylib/boututils/datafile.py:
```diff
@@ -319,9 +319,13 @@ def __getitem__(self, name):
 
+        These are attributes for varname if varname is specified. If varname is
+        the empty string (default case) then the global attributes of the file
+        are returned.
```
Does this actually work? I can't see anything that handles the empty string in the NetCDF/HDF5 implementations of `attributes`. It looks like they both try and find attributes for a variable named `""`, i.e. the empty string. As there likely *isn't* a variable called this, they return `None`. I might be wrong, I don't have a file with file-level attributes to hand!

Also, I'm assuming "global attributes" means "file attributes"?

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -554,10 +557,15 @@ def __init__(self, path=".", prefix="BOUT.dmp", suffix=None, caching=False,
         self.dimensions = {}
-        self.evolvingVariableNames = []
+        self.evolvingVariables = []
+        self._shiftAngle = None # used for getShifted()/getUnshifted() methods, calculated from self['zShift'] when first needed
```
Please put long comments like this on separate lines

ZedThree commented on tools/pylib/boutdata/data.py:
```diff
@@ -599,11 +607,47 @@ def keys(self):
+        """
+        Return a variable shifted by zShift: if BOUT++ output is in
+        'unshifted', i.e. toroidal, coordinates, this function returns a field
+        in field-aligned coordinates
```
This sounds a bit like it checks if the output file is in unshifted space, and always does the "correct" thing, but if I understand correctly, it unconditionally shifts the variable?

Always doing the "correct thing" would be best; if that's not possible currently, please could you change the docstring to say no checking is done (or something along those lines)

johnomotani commented:
@ZedThree thanks for the comments!

The parts using file attributes to read options I should never have put in here. They were meant to go with another project which is incomplete anyway. Sorry! The support for file attributes was indeed broken, but I've fixed and tested it now. I think it may be useful in future.

To allow getShifted/getUnshifted to do the right thing automatically, I've added another attribute to keep track of whether Field3Ds are in orthogonal or field-aligned coordinates. It required some changes to the BOUT++ code as well, so suggestions on a simpler way are welcome, but what I've added seems useful anyway. Don't know if we're ever likely to handle fields on a mix of orthogonal and field-aligned coordinates, but if we do there's some support for keeping track of it. Possibly coordinate_system should be an enum like CELL_LOC rather than a string?

johnomotani commented:
This PR would resolve #46.

johnomotani commented:
This PR should be re-written to use the `coordinate_system` attribute introduced in #1459 when that PR is merged.

johnomotani commented:
To resolve #46, the Python code in this PR could be modified easily to recognise the `direction_y` attribute instead of `coordinate_system` (to check whether a variable is in field-aligned or toroidal coordinates), some naming updated (e.g. `toFieldAligned` and `fromFieldAligned` instead of `getShifted` and `getUnshifted`) and the C++ changes removed. But since the feature is already provided and tested in `xBOUT`, I'd suggest just closing this PR.

ZedThree commented:
Ok, closing as xBOUT replaces this functionality
